tests/language-feature-scripts/lazy-loading-__destruct-throws.phpt

```diff
@@ -0,0 +1,26 @@
+--TEST--
+Verifies that generated lazy loading value holders disallow protected property direct unset
+--FILE--
+<?php
+
+require_once __DIR__ . '/init.php';
+
+class Destructable
+{
+    public function __destruct()
+    {
+        # code...
+    }
+}
+
+$factory = new \ProxyManager\Factory\LazyLoadingValueHolderFactory($configuration);
+
+$proxy = $factory->createProxy(Destructable::class, function (& $wrapped, $proxy, $method, array $parameters, & $initializer) {
+    $initializer = null;
```

Comment:
Maybe this should throw to reproduce the issue?

Reply:
I will have to ask @mnapoli about how container proxies works, because it only happens with this option enabled. 🤔

Reply:
Hi, I'm not familiar with the issue here sorry. But I don't see why an exception should be thrown, especially if the test case reproduces the issue.

Reply:
@odahcam the issue is not reproduced here:

```
1) /home/travis/build/Ocramius/ProxyManager/tests/language-feature-scripts/lazy-loading-__destruct-throws.phpt
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
-%SUncaught Error:%sCall to a member function __destruct() on null in %a
/home/travis/build/Ocramius/ProxyManager/tests/language-feature-scripts/lazy-loading-__destruct-throws.phpt:26
```

@mnapoli I'm also having trouble in understanding the full scenario for this issue, sorry...

Reply:
Yesterday I started to think this is a issue with PHP-DI, since the problem only occours when I enable the option `$container->writeProxiesToFiles(true, 'folder/path')`. If @mnapoli could explain what that method does I would understand it better.

@Ocramius maybe here https://github.com/PHP-DI/PHP-DI/issues/669 you can find more information.

At the end I don't know what exactly is happening either, I'm just facing the problem, where one of the proxies generated by PHP-DI using this library throws inside a `__destruct` method because the instance of the class the proxy is for is null inside the proxy class. Yes confusing and strange.

As you can see, things are null here:

![image](https://user-images.githubusercontent.com/3942006/60399821-32dc7900-9b41-11e9-8d0e-cb06403863fe.png)

And then the error occours:

![image](https://user-images.githubusercontent.com/3942006/60399831-5b647300-9b41-11e9-88d5-f3eae043c862.png)

Maybe you understand better than me whats happening.

Reply:
What's the initializer there? It got called, but not reset to `null`, which to me smells like it threw an exception.

Reply:
```
<?php
$this->initializer5d18ea48bcba2345610290 && $this->initializer5d18ea48bcba2345610290->__invoke($this->valueHolder5d18ea48bcb97124004552, $this, '__destruct', array(), $this->initializer5d18ea48bcba2345610290);
```

![image](https://user-images.githubusercontent.com/3942006/60405590-7e197a80-9b87-11e9-94a7-70fc01256ab0.png)

Is there a way for me to check if it throwed?

Reply:
I don't store such information, but I'm pretty sure that the engine will aggressively call `__destruct()` anyway.

I'd use a step debugger and put breakpoints on any of the API of the generated code.